framework/src/iotbus/iotbus_gpio.c

```diff
@@ -400,88 +416,145 @@ int iotbus_gpio_unregister_cb(iotbus_gpio_context_h dev)
 	struct _iotbus_gpio_s *handle;
-	struct iotbus_int_info_s info = { 0, };
-
-	handle = (struct _iotbus_gpio_s *)dev->handle;
-	handle->cb = cb;
-
-	info.handle = dev;
-	info.pin_type = IOTBUS_GPIO;
-	info.int_type = int_type;
-	info.pid = iotapi_get_pid();
 	
-	fd = open(IOTBUS_SIGPATH, O_NONBLOCK);
-	if (fd < 0) {
-		ret = get_errno();
-		ibdbg("Open iotsig fail[%d, %d]\n", fd, ret);
-		return IOTBUS_ERROR_DEVICE_FAIL;
+	int ret;
+	siginfo_t info;
+	sigset_t sig_set;
+	bool running = true;
+	void *ptr;
+
+	sigemptyset(&sig_set);
+	sigaddset(&sig_set, SIG_IOTBUS_GPIO_RISING);
+	sigaddset(&sig_set, SIG_IOTBUS_GPIO_FALLING);
+
+	pthread_sigmask(SIG_BLOCK, &sig_set, NULL);
+
+	while (running) {
+		ibdbg("Waiting GPIO Interrupt Signal...\n");
+		ret = sigwaitinfo(&sig_set, &info);
+#ifdef CONFIG_CAN_PASS_STRUCTS
+		ptr = info.si_value.sival_ptr;
+#else
+		ptr = info.si_value;
+#endif
+		if (ptr == NULL) {
+			ibdbg("Handler is not assigned.\n");
+			continue;
+		}
+		dev = (iotbus_gpio_context_h)ptr;
+		handle = (struct _iotbus_gpio_s *)dev->handle;
+		if (ret == SIG_IOTBUS_GPIO_RISING) {
+			ibdbg("Get GPIO Rising Interrupt in [%d]\n", handle->pin);
+		} else if (ret == SIG_IOTBUS_GPIO_FALLING) {
+			ibdbg("Get GPIO Falling Interrupt in [%d]\n", handle->pin);
+		}
+		// Call callback function.
+		((iotbus_gpio_cb)handle->cb)(dev);
 	}
+	
+	pthread_sigmask(SIG_UNBLOCK, &sig_set, NULL);
+	pthread_exit(NULL);
+
+	return NULL;
+}
+
+int iotbus_gpio_set_interrupt(iotbus_gpio_context_h dev, iotbus_gpio_edge_e int_type, iotbus_gpio_cb cb)
+{
+	int ret = -1;
+	struct _iotbus_gpio_s *handle;
+	FAR struct gpio_notify_s notify;
 
-	if (int_type == IOTBUS_GPIO_FALLING) {
-		edge = GPIO_EDGE_FALLING;
-	} else if (int_type == IOTBUS_GPIO_EDGE_RISING) {
-		edge = GPIO_EDGE_RISING;
-	} else {
-		edge = GPIO_EDGE_NONE;
+	if (!dev || !dev->handle || int_type == IOTBUS_GPIO_EDGE_NONE) {
+		return IOTBUS_ERROR_INVALID_PARAMETER;
 	}
 
-	ret = ioctl(fd, IOTBUS_INTR_REGISTER, (unsigned long)&info);
-	if (ret < 0) {
-		ibdbg("Fail to register %d\n", int_type);
-		close(fd);
-		return IOTBUS_ERROR_DEVICE_FAIL;
+	handle = (struct _iotbus_gpio_s *)dev->handle;
+
+	switch (int_type) {
+	case IOTBUS_GPIO_EDGE_BOTH:
+		return IOTBUS_ERROR_NOT_SUPPORTED;
+	case IOTBUS_GPIO_EDGE_RISING:
+		notify.gn_rising  = true;
+		notify.gn_falling = false;
+		notify.gn_signo = SIG_IOTBUS_GPIO_RISING;
+		break;
+	case IOTBUS_GPIO_EDGE_FALLING:
+		notify.gn_rising  = false;
+		notify.gn_falling = true;
+		notify.gn_signo = SIG_IOTBUS_GPIO_FALLING;
+		break;
+	default:
+		return IOTBUS_ERROR_INVALID_PARAMETER;
 	}
-	ret = ioctl(handle->fd, GPIOIOC_SET_INTERRUPT, edge);
-	if (ret < 0) {
-		ibdbg("Fail to set %d\n", int_type);
-		close(fd);
-		return IOTBUS_ERROR_DEVICE_FAIL;
+
+	struct intr_attr val = { CONFIG_IOTBUS_GPIO_INTERRUPT_PRIORITY, IOTBUS_GPIO, getpid() };
+
+	handle->pid = create_intr_pthread(gpio_intr_thread, (void *)dev, &val);
```

Comment:
You can create only one thread like a singleton method, because all request uses just one device, finally.
Please think about more about management of callbacks from applications..